uint32 xpos
uint32 ypos
